music_assistant/providers/lastfm/__init__.py

```diff
@@ -0,0 +1,304 @@
+            return
+
+        self._network = _get_network(self._get_network_config())
+
+        # subscribe to internal event
+        self._on_unload.append(
+            self.mass.subscribe(self._on_mass_media_item_played, EventType.MEDIA_ITEM_PLAYED)
+        )
+
+    async def unload(self, is_removed: bool = False) -> None:
+        """
+        Handle unload/close of the provider.
+
+        Called when provider is deregistered (e.g. MA exiting or config reloading).
+        """
+        for unload_cb in self._on_unload:
+            unload_cb()
+
+    async def _on_mass_media_item_played(self, event: MassEvent) -> None:
+        """Media item has finished playing, we'll scrobble the track."""
+        if self._network is None:
+            self.logger.error("no network available during _on_mass_media_item_played")
+            return
+
+        report = event.data
+
+        def update_now_playing() -> None:
+            try:
+                self._network.update_now_playing(
+                    report.artist,
+                    report.name,
+                    report.album,
+                    duration=report.duration,
+                    mbid=report.mbid,
+                )
+                self.logger.debug(f"track {report.uri} marked as 'now playing'")
+                self._currently_playing = report.uri
+            except Exception as err:
+                self.logger.exception(err)
+
+        def scrobble() -> None:
+            try:
+                # album artist and track number are not available without an extra API call
+                # so they won't be scrobbled
+                self._network.scrobble(
+                    report.artist,
+                    report.name,
+                    time.time(),
+                    report.album,
+                    duration=report.duration,
+                    mbid=report.mbid,
+                )
+            except Exception as err:
+                self.logger.exception(err)
+
+        # update now playing if needed
+        if self._currently_playing is None or self._currently_playing != report.uri:
+            self.mass.loop.run_in_executor(None, update_now_playing)
+
+        if self.should_scrobble(report):
+            self.mass.loop.run_in_executor(None, scrobble)
+
+        if report.fully_played:
+            # reset currently playing to avoid it expiring when looping songs
+            self._currently_playing = None
+
+    def should_scrobble(self, report: MediaItemPlaybackProgressReport) -> bool:
+        """Determine if a track should be scrobbled, to be extended later."""
+        # ideally we want more precise control
+        # but because the event is triggered every 30s
+        # and we don't have full queue details to determine
+        # the exact context in which the event was fired
+        # we can only rely on fully_played for now
+        return bool(report.fully_played)
```

Comment:
you also have a "is_playing" attribute on the event now.
So you could consider the following logic:

scrobble if playing and progress >= 50%
or
scrobble if fully played 

In the future we can also replace fully played and is_playing booleans with some enum value of specific state

Reply:
oh nice, that does help. Although I'm not sure what the direct implications are, as I then need to keep state to not double trigger a scrobble. Which becomes interesting when considering single songs being on loop (which I personally do quite a lot 😂), as it's hard to detect when it restarts. 

I tried to fix it when I looked at the edge-case I mentioned, but it's not very straightforward currently, so I didn't add it to the MR. But it would be nice if the playback report had an indication that the event specifically indicates start of queued item and end of queued item.

Reply:
OK, we can extend/iterate on it later once this is in and we have some feedback and actually daily use. 
Not sure if you're on discord but we could have a chat there how to best move forward on this.

Reply:
Yeah I am, same username, although not really active besides in one of the LastFM threads 😂